backup-tools/backup-replication-image/go-src/job_queue.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	_ "expvar"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ncw/swift"
+	"github.com/sapcc/containers/backup-tools/go-src/configuration"
+	"github.com/sapcc/containers/backup-tools/go-src/swiftcli"
+	"github.com/sapcc/containers/backup-tools/go-src/utils"
+)
+
+const maxWorkers = 5 // normal 5; debug 2
+
+var (
+	alreadyPrinted   int
+	currentFilesDone = 0
+	expiration       int64
+	backupDir        = "/backup/tmp"
+	EnvFrom          *Env
+	EnvTo            = make([]*Env, 2)
+)
+
+type Env struct {
+	Cfg      configuration.EnvironmentStruct
+	Files    []string
+	SwiftCli *swift.Connection
+}
+
+type Job struct {
+	EnvFrom      *Env
+	EnvTo        []*Env
+	File         string
+	FileAllCount int
+	FileNumber   int
+}
+
+func doWork(id int, j Job) {
+	var dlErr, upErr error
+	var uploadDone bool
+	var dlFilePath string
+	lastDownload := false
+	from := j.EnvFrom
+
+	if DebugOutput == "yes" {
+		log.Printf("Worker%d File %d/%d: started process file %s\n", id, j.FileNumber, j.FileAllCount, j.File)
+	}
+	for _, to := range j.EnvTo {
+		uploadDone = false
+
+		// skip if file already on the replication region present
+		if stringInSlice(j.File, to.Files) {
+			if DebugOutput == "yes" {
+				log.Printf("Worker%d File %d/%d: Skip %s to %s\n", id, j.FileNumber, j.FileAllCount, j.File, to.Cfg.OsRegionName)
+			}
+			continue
+		}
+
+		// Download only file if not already done
+		if !lastDownload {
+			if DebugOutput == "yes" {
+				log.Printf("Worker%d File %d/%d: Download File: %s from %s\n", id, j.FileNumber, j.FileAllCount, j.File, from.Cfg.OsRegionName)
+			}
+			dlFilePath, dlErr = swiftcli.SwiftDownloadFile(from.SwiftCli, j.File, &backupDir, true)
+			if dlErr != nil {
+				PromGauge.SetError()
+				log.Printf("Worker%d File %d/%d (%s): Download File Error: %s\n", id, j.FileNumber, j.FileAllCount, j.File, dlErr)
+				// GoToEndWhileDownloadError used for exit the replication for error while downloading the file
+				goto GoToEndWhileDownloadError
+				break
+			}
+			lastDownload = true
+		}
+
+		// Upload file if no error before was triggered, and the dlFilePath is longer then 0
+		if dlErr == nil && len(dlFilePath) > 0 {
+			fakeName := strings.TrimPrefix(dlFilePath, backupDir)
+			if DebugOutput == "yes" {
+				log.Printf("Worker%d File %d/%d: Upload File: %s to %s\n", id, j.FileNumber, j.FileAllCount, fakeName, to.Cfg.OsRegionName)
+			}
+			uploadDone, upErr = swiftcli.SwiftUploadFile(to.SwiftCli, dlFilePath, &expiration, &fakeName)
+			if upErr != nil || !uploadDone {
+				PromGauge.SetError()
+				log.Printf("Worker%d File %d/%d (%s): Upload File Error: %s\n", id, j.FileNumber, j.FileAllCount, j.File, upErr)
+			}
+		}
+	}
+
+	// GoToEndWhileDownloadError used for exit the replication for error while downloading the file
+GoToEndWhileDownloadError:
```

Comment:
You are always going here (there's no `return` before), so I'm pretty sure there's something wrong here in the success case. In general, please avoid `goto`.

Reply:
This is only in case the download was not success so you will for sure always going to this Mark but you over jump the upload of a not existing file.